### PR TITLE
Java: Model `SerialKiller`

### PR DESCRIPTION
**Pull request by jorgectf**
This PR introduces a check for `SerialKiller`, a library used to prevent deserialization of arbitrary classes.

### CODE REVIEW

**Comment by pwntester** [comment]
@jorgectf @atorralba The `resolveClass` main purpose is not security. Marking a subclass of ObjectInputStream as safe because it overrides the `resolveClass` method may lead to many FN. There are also some look-ahead objectInputStreams that may be applying block lists instead of allow lists and which may allow dangerous gadgets.

File: java/ql/test/query-tests/security/CWE-502/A.java [hunk 1]
```diff
@@ -21,20 +21,37 @@ public Object deserialize2(Socket sock) throws java.io.IOException, ClassNotFoun
     return in.readUnshared(); // $unsafeDeserialization
   }
 
-  public Object deserialize3(Socket sock) throws java.io.IOException {
+  public Object deserialize3(Socket sock) throws java.io.IOException, ClassNotFoundException {
```
Comment (atorralba) [inline java/ql/test/query-tests/security/CWE-502/A.java:24]: Should we test unsafe `resolveClass` implementations as well? Something like:

```java
protected Class<?> resolveClass(ObjectStreamClass osc) throws java.io.IOException, ClassNotFoundException {
  return super.resolveClass(osc);
}
```

**Review by atorralba — COMMENTED** [review]
This mostly LGTM, thanks for the contribution! I added a couple of comments:

**Comment by jorgectf** [comment]
@pwntester @atorralba: After reviewing the use of `resolveClass` across OSS:

I see `consider it a sanitizer if it's not always returning a call to resolveClass` very unreliable, even if we used dataflow. I would change to checking if `resolveClass[0]` flows to a guard like `equals()`, `startswith()` or `contains()`.

We should also consider modeling [SerialKiller](https://github.com/ikkisoft/SerialKiller).

All of this would remove a few FPs from our lists.

File: java/ql/test/query-tests/security/CWE-502/A.java (hunk 1 above)
Comment (jorgectf) [reply java/ql/test/query-tests/security/CWE-502/A.java:24]: Nice catch! Thanks! https://github.com/github/codeql/pull/13358/commits/5ea5ca42a4d124897c0d1f2a0398ad74534d69e7

**Comment by jorgectf** [comment]
I've come to the conclusion that modeling `resolveClass` is not worth it. I've left the model for `SerialKiller`, and so I'm marking this PR as ready for review.

File: java/ql/lib/semmle/code/java/security/UnsafeDeserializationQuery.qll [hunk 2]
```diff
@@ -28,6 +28,20 @@ private class ObjectInputStreamReadObjectMethod extends Method {
   }
 }
 
+/**
+ * A type coming from `ObjectInputStream` that makes it safe to deserialize untrusted data.
```
Comment (atorralba) [inline java/ql/lib/semmle/code/java/security/UnsafeDeserializationQuery.qll:32]:
```suggestion
 * A type extending `ObjectInputStream` that makes it safe to deserialize untrusted data.
```

**Review by atorralba — APPROVED** [review]
This looks much less controversial :)

**Comment by Sugarpie86** [comment]
> @jorgectf @atorralba The `resolveClass` main purpose is not security. Marking a subclass of ObjectInputStream as safe because it overrides the `resolveClass` method may lead to many FN. There are also some look-ahead objectInputStreams that may be applying block lists instead of allow lists and which may allow dangerous gadgets.
> 
> We can run the `SafeObjectInputStreamResolveClassMethod` on 1000 repos using OIS and check the ration of effectiveness.

**Comment by atorralba** [comment]
DCA looks good too, merging.

**Pull request merged**